lib/services/local/identitycenter_test.go

```diff
@@ -0,0 +1,282 @@
+
+				// GIVEN an IdentityCenter service populated with a resource
+				uut, err := NewIdentityCenterService(IdentityCenterServiceConfig{Backend: backend})
+				require.NoError(t, err)
+				createdResource := test.createResource(t, ctx, uut, resourceID)
+
+				// WHEN I modify the backend record for that resource...
+				tmpResource, err := test.getResource(ctx, uut, resourceID)
+				require.NoError(t, err)
+				tmpResource.GetMetadata().Labels = map[string]string{"update": "1"}
+				updatedResource, err := test.updateResource(ctx, uut, tmpResource)
+				require.NoError(t, err)
+
+				// EXPECT that any attempt to update backend via the original in-memory
+				// version of the resource fails with a comparison error
+				createdResource.GetMetadata().Labels = map[string]string{"update": "2"}
+				_, err = test.updateResource(ctx, uut, createdResource)
+				require.True(t, trace.IsCompareFailed(err), "expected a compare failed error, got %T (%s)", err, err)
+
+				// EXPECT that the backend still reflects the first updated revision,
+				// rather than failed update
+				r, err := test.getResource(ctx, uut, resourceID)
+				require.NoError(t, err)
+				require.Equal(t, "1", r.GetMetadata().Labels["update"])
+
+				// WHEN I attempt update the backend via the latest revision of the
+				// record...
+				updatedResource.GetMetadata().Labels["update"] = "3"
+				_, err = test.updateResource(ctx, uut, updatedResource)
+
+				// EXPECT the update to succeed, and the backend record to have been
+				// updated
+				require.NoError(t, err)
+				r, err = test.getResource(ctx, uut, resourceID)
+				require.NoError(t, err)
+				require.Equal(t, "3", r.GetMetadata().Labels["update"])
+			})
+
+			t.Run("UnconditionalUpsert", func(t *testing.T) {
+				t.Parallel()
+
+				if test.upsertResource == nil {
+					t.Skip(test.name + " does not support unconditional upsert")
+				}
+
+				ctx := newTestContext(t)
+				clock := clockwork.NewFakeClock()
+				backend := newTestBackend(t, ctx, clock)
+
+				// GIVEN an IdentityCenter service populated with a resource
+				uut, err := NewIdentityCenterService(IdentityCenterServiceConfig{Backend: backend})
+				require.NoError(t, err)
+				createdResource := test.createResource(t, ctx, uut, resourceID)
+
+				// WHEN I modify the backend record for that resource...
+				tmpResource, err := test.getResource(ctx, uut, resourceID)
+				require.NoError(t, err)
+				tmpResource.GetMetadata().Labels = map[string]string{"update": "1"}
+				_, err = test.updateResource(ctx, uut, tmpResource)
+				require.NoError(t, err)
+
+				// EXPECT that an upsert will overwrite the modified record, even
+				// if the revision source revision of the change is incorrect
+				createdResource.GetMetadata().Labels = map[string]string{"update": "2"}
+				_, err = test.upsertResource(ctx, uut, createdResource)
+				require.NoError(t, err)
+
+				// EXPECT that the backend reflects the updated revision
```

Comment:
nit: I would test with a new resource value instead of reusing `createdResource` here just to make it clear that upsert is actually using entirely new values. 

I was also initially confused with the usage of `revision` because we have a dedicated `revision` field in the resource metadata header but below we just assert matching labels.